File: 79. Word Search new.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        visited = [[False for j in xrange(len(board[0]))] for i in xrange(len(board))]
        for i in xrange(len(board)):
            for j in xrange(len(board[0])):
                if self.DFS(i, j, 0, visited, board, word):
                    return True
        return False
    
    def DFS(self, i, j, windex, visited, board, word):
        if windex == len(word):
            return True    
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or visited[i][j] or board[i][j] != word[windex]:
            return False
        visited[i][j] = True
        res = self.DFS(i-1, j, windex+1, visited, board, word) or self.DFS(i+1, j, windex+1, visited, board, word) or self.DFS(i, j-1, windex+1, visited, board, word) or self.DFS(i, j+1, windex+1, visited, board, word)
        visited[i][j] = False
        return res

File: 79. Word Search new.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        vector <vector<bool>> visited (board.size(), vector <bool> (board[0].size(), false));
        for (int i = 0; i < board.size(); i++)
        {
            for (int j = 0; j < board[0].size(); j++)
            {
                if (DFS(board, i, j, 0, word))
                    return true;
            }
        }
        return false;
    }
    
    bool DFS(vector<vector<char> > &board, int i, int j, int windex, string word)
     {
         if(board[i][j] == word[windex])
         {
             if(windex == word.size() - 1)
                 return true;
             board[i][j] = '*';
             if(i + 1 < board.size() && DFS(board, i + 1, j, windex + 1, word))
                 return true;
             if(i - 1 >= 0 && DFS(board, i - 1, j, windex + 1, word))
                 return true;
             if(j + 1 < board[0].size() && DFS(board, i, j + 1, windex + 1, word))
                 return true;
             if(j - 1 >= 0 && DFS(board, i, j - 1, windex + 1, word))
                 return true;
             board[i][j] = word[windex];
         }
         return false;
     }
};
